### ISSUE TITLE
[cssom] Add `toString` method on `CSSStyleSheet`

### ISSUE THREAD

**Issue by lucacasonato** [w3c/csswg-drafts#7171]
CSSOM allows programmatic construction and modification of `CSSStyleSheet`s. These created sheets can be attached to a HTML document on the client so users can directly use them. It is however not currently trivial to "export" programmatic style sheets back into a text representation.

This is unfortunate, because the CSSOM API would be great for the generation of style sheets server side, if the server side runtime were to support the API. This is something that we would like to do in Deno (https://github.com/denoland/deno/issues/13898). In Deno, there is no DOM, so the style sheet could never directly be used. Instead you would build a style sheet with the CSSOM API, and then export it to a string and send it to a client.

It is possible to turn a `CSSStyleSheet` into a string my manually concatenating the string representations of each rule. This is what the [`stringify-css-stylesheet` npm package](https://github.com/tomhodgins/stringify-css-stylesheet/blob/master/index.js) does for example. A simplified example:

```js
function stringifyStyleSheet(stylesheet) {
  return Array.from(stylesheet.cssRules)
    .map(rule => rule.cssText || "")
    .join("\n")
}
```

It would be awesome if `CSSStyleSheet.prototype.toString()` would exist to do this natively, without requiring any helper code.

**Comment by tabatkins** [comment]
Yes, this seems like a lack purely due to apathy, not due to any actual problems, since authors can *trivially* serialize an entire sheet anyway, as demonstrated. I think this would be completely fine to add.

I couldn't find any discussions about this in the past with a quick search thru my archives, but that doesn't guarantee there wasn't any. Is anyone aware of us discussing this in the past?

**Comment by timotius02** [comment]
I'm currently taking this issue after some guidance from @tabatkins (thanks a lot for your guidance btw). 

I wanted to know whether the stringifier should be the toString method of CSSStyleSheet like this issue suggests or if it should be exposed as a string property just like in CSSRule. Would love some opinions on this.

**Comment by tabatkins** [comment]
For consistency, we probably should expose it as `cssText`, but we can make it better by making it a [stringifier attribute](https://webidl.spec.whatwg.org/#idl-stringifiers).

**Comment by Loirooriol** [comment]
`CSSStyleRule` doesn't have a stringifier, so I would keep `CSSStyleSheet` consistent.

**Comment by lucacasonato** [comment]
Or we add a stringifier to both?

**Comment by timotius02** [comment]
I would  be open to changing both, but what are the chances that there's something out there that relies on the current stringified behavior of CSSStyleRule?

Also is there a difference spec-wise other than adding stringifier to the attribute to make something a stringifier?

**Comment by tabatkins** [comment]
Nope, that keyword is all you need.

(If you don't have an attribute already containing the text you have to write the definition *slightly* different to have an independent stringifier, but it's literally just a few words difference in invocation.)

And yeah, adding a stringifier to both would be ideal (but it increases the testing cost slightly). Relying on the current stringification  seems not too likely. (In theory someone could be using it for dirty type-testing since it currently stringifies to `"[Object CSSRule]"`, but doing that generically requires you to specifically use `Object.prototype.toString.call()` precisely *because* a lot of objects have stringifiers, so it's likely such uses are relying on that and won't be affected.)

**Comment by Loirooriol** [comment]
It should also be consistent with other interfaces like `CSSStyleDeclaration`. And given that most things don't have stringifiers, I wouldn't add them here.

**Comment by timotius02** [comment]
On the other side however, it seems like MediaList interface has a stringifier, though it seems like it's the only one in cssom and I can't seem to track the reasoning behind that. 

We can choose add a stringifier to CSSStyleDeclaration as well but if we do want to move in the path of least resistance, just forgoing stringifiers is best. Can anyone provide context on the state of stringifiers as a whole in w3c spec, like how common are they currently and any specs that heavily uses stringifiers as major features?

**Comment by trusktr** [comment]
Here's a use case for this:

- https://github.com/WICG/webcomponents/issues/909#issuecomment-1793327671

In the second snippet there, the `#shared-stylesheet` could have a CSS OM and no `textContent`, so to-stringing it would be handy.

**Comment by TheJaredWilcurt** [comment]
**Use case:** Visual Regression (screenshot) testing

There is a problem with visual regression tests where they aren't cross-browser or cross-platform compatible due to subtle differences in rendering, especially fonts. So taking a screenshot of the app in a specific state, and then comparing the pixels against a previous known-good screenshot will fail just because the screenshot was done on a different OS or slightly different browser version.

If you could serialize the entire DOM and CSSOM and get hashes of each, then store those hashes next to the screenshot, you'd be able to inform a CI tool to update the snapshots when it re-runs if the hashes do not match. Allowing you to locally update the hashes when you want a new known good, without having to commit a screenshot from your machine that will only work for your machine. Then when the CI runs, it validates the screenshots, or replaces them if the hashes don't match.

This would work best if you could serialize just the part of the DOM and CSSOM that applies to a specific selector, like `[data-test="myWidget"]`. So the screenshot does not need to be of the entire page, but just the part related to that specific test.

**Comment by mayank99** [comment]
Ran into a super weird issue today when using the `cssText` shown in the original description. Apparently it can produce invalid CSS when using custom properties: https://issues.chromium.org/issues/40804066

Obviously the bug should be fixed, but a `.toString()` function could help avoid the problem altogether.

**Comment by davejsdev** [comment]
Another use case for this:

Tools that generate CSS output would benefit from being able to programmatically build and output CSS text using a standard API that ensures correctness. We're doing this in [Terrazzo's CSS plugin](https://github.com/terrazzoapp/terrazzo/issues/166) where we generate custom properties from Design Tokens. This would be a lot easier and reliable if we could reliably build valid CSS with the CSSOM spec and then print the output to a file.

A similar use case to this is exporting from platforms like Webflow.

**Comment by cdoublev** [comment]
Would it be allowed for **cross-origin** `@import`-ed style sheets? (Access to `CSSStyleSheet.cssRules` is forbidden for them.)
